iwyu_preprocessor.cc

```diff
@@ -1087,4 +1129,25 @@ bool IwyuPreprocessorInfo::ForwardDeclareIsInhibited(
       ContainsKey(*inhibited_forward_declares, normalized_symbol_name);
 }
 
+bool IwyuPreprocessorInfo::ForwardDeclareInKeepRange(
+    const SourceLocation& loc) const {
+  // Check if the entry exists, shortcircuit if not
+  const FileEntry* file = GetFileEntry(loc);
+
+  auto keep_ranges = keep_location_ranges_.equal_range(file);
+
+  // with multimaps first==second if and only if there's no matches
+  if (keep_ranges.first == keep_ranges.second) {
+    return false;
+  }
+
+  const SourceRange decl_range{loc};
+  for (auto it = keep_ranges.first; it != keep_ranges.second; ++it) {
```

Comment:
Looking at this loop now, I notice a few things --

* the short-circuit can be removed, it's the automatic behavior of the loop
* `SourceRange` implicitly converts from `SourceLocation`, so you don't need `decl_range`

So I think you could get away with:
```
const FileEntry* file = GetFileEntry(loc);
auto keep_ranges = keep_location_ranges_.equal_range(file);
for (auto it = keep_ranges.first; it != keep_ranges.second; ++it) {
  if (it->second.fullyContains(loc) {
    return true;
  }
}
return false;
```